### PR TITLE
Add an option to skip importing root namespace classes (like \DateTime)

### PR DESCRIPTION
**Pull request by gnutix**
Trying to fix https://github.com/rectorphp/rector/issues/1911#issuecomment-525642537

```yaml
services:
    Rector\CodingStyle\Rector\Namespace_\ImportFullyQualifiedNamesRector:
        $shouldImportRootNamespaceClasses: false
```

I still need to prevent changing DocBlock annotations like `@var \DateTime`... which seems way harder, looking at the code.

### CODE REVIEW

**Comment by TomasVotruba** [comment]
Great pick!

Definitely start with a test with optional parameter to be on

**Comment by gnutix** [comment]
Test added. But some help to get the condition `$node is NonRootNamespace` right would be more than welcome. Currently my condition broke the tests in `packages/CodingStyle/tests/Rector/Namespace_/ImportFullyQualifiedNamesRector/NonNamespacedTest.php`. And I've had a look at the DocBlock issue, and the code is quite complex in there (`DocBlockManipulator`, `DocBlockNameImporter`, `PhpDocNodeTraverser`, etc); I wouldn't know how/where to pass that new option around and where to use it exactly so I wouldn't break half the project.

**Comment by TomasVotruba** [comment]
> But some help to get the condition $node is NonRootNamespace right would be more than welcome.

From top of my head, I'd detect `\DateTime` like:

```
$name = $this->getName($node);
if (Strings::startsWith($name, '\\')) {
   if (substr_count($name, '\\') === 1) {
        // skip
	    return null;
   }
}
```

**Comment by gnutix** [comment]
Here's a working implementation. I've not handled DocBlock though, so the test is still failing.

File: packages/CodingStyle/src/Rector/Namespace_/ImportFullyQualifiedNamesRector.php [hunk 1]
```diff
@@ -72,6 +81,14 @@ public function getNodeTypes(): array
...
+        if (! $this->shouldImportRootNamespaceClasses && $node instanceof Name) {
+            $name = $this->getName($node);
+            if ($name !== null && substr_count($name, '\\') === 0) {
+                return null;
+            }
+        }
+
         $this->useAddingCommander->analyseFileInfoUseStatements($node);
 
         if ($node instanceof Name) {
```
Comment (TomasVotruba) [inline packages/CodingStyle/src/Rector/Namespace_/ImportFullyQualifiedNamesRector.php]: Why not add it here, where `Name` check is already added?

**Comment by TomasVotruba** [comment]
As for doblocks, this might be place to check it:
https://github.com/rectorphp/rector/blob/master/packages/NodeTypeResolver/src/PhpDoc/NodeAnalyzer/DocBlockNameImporter.php#L125

**Comment by gnutix** [comment]
I've finished the implementation. :+1:

**Comment by TomasVotruba** [comment]
Feature looks ready, good job :+1: 

> Would there be some documentation to write for this new option ?

The documentation is generated from code samples, so adding one `ConfiguredCodeSample()` in `getDefinition()` will be enough

**Comment by TomasVotruba** [comment]
> I've finished the implementation. +1 Would there be some documentation to write for this new option ?

Feel free to send that in next PR. I'm merging this, since feature is complete

**Comment by gnutix** [comment]
Not sure how I can describe the new parameter within `new CodeSample(...)` as this only provides before/after code examples - but not the constructor parameters involved to achieve such results.

**Comment by TomasVotruba** [comment]
`ConfiguredCodeSample`

**Comment by gnutix** [comment]
Sorry, I should have read more attentively! :roll_eyes:

**Comment by TomasVotruba** [comment]
No troubles :+1:

**Comment by gnutix** [comment]
@TomasVotruba I can't find this option in `master` anymore. Was it removed purposefully ? :thinking:

**Comment by TomasVotruba** [comment]
Could you add that to README as well?

**Comment by Aerendir** [comment]
@TomasVotruba I documented it

**Pull request merged**